### PR TITLE
Shader Stub Code, Clipping Space Unification and GCC Warnings

### PR DESCRIPTION
**Pull request by devshgraphicsprogramming**
@Crisspl status update on the number of GCC bugs?

### CODE REVIEW

File: 3rdparty/CMakeLists.txt [hunk 1]
```diff
@@ -12,24 +12,34 @@ OPTION(BUILD_TESTS ON)
...
-	libpng/pngwio.c
-	libpng/pngwrite.c
-	libpng/pngwtran.c
-	libpng/pngwutil.c
-)
-target_include_directories(libpng PUBLIC zlib)
+#manipulating libpng cmake options from within cmake script
+option(PNG_BUILD_ZLIB "Custom zlib Location, else find_package is used" OFF)
+set(PNG_BUILD_ZLIB ON) #pnglib depends on zlib and zlib is part of Irrlicht
+set(ZLIB_INCLUDE_DIR ${THIRD_PARTY_SOURCE_DIR}/zlib) #for pnglib
```
Comment (devshgraphicsprogramming) [inline 3rdparty/CMakeLists.txt:20]: I trust you this works ;)

**Comment by devshgraphicsprogramming** [comment]
@Crisspl I think you've missed out on aesGladman
https://github.com/BrianGladman/aes

and furthermore does the CMakeLists script automatically update the submodules + pull the latest version of OpenSSL 1.0.2 ?

**Comment by Crisspl** [comment]
> I think you've missed out on aesGladman

see commit message

> furthermore does the CMakeLists script automatically update the submodules + pull the latest version of OpenSSL 1.0.2 ?

just working on it, should be today

**Comment by devshgraphicsprogramming** [comment]
> see commit message

I understand.

> just working on it, should be today

Regarding, OpenSSL can you do `#ifdef` to detect version 1.0.2 and make the code work both with 1.0.2 and 1.1.1 ?

The ver 1.0.2 branch should still be our default submodule.

File: 3rdparty/openssl/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+set(NMAKE_64_EXECUTABLE NOT_FOUND CACHE FILEPATH "Path to x64 nmake.exe")
+
+set(VS140_SCRIPTS_DIR $ENV{VS140COMNTOOLS})
+
+add_custom_target(
+  openssl_build ALL
+  #COMMAND "${VS140_SCRIPTS_DIR}vsvars32.bat"
+  COMMAND "${PERL_EXECUTABLE} ./Configure VC-WIN32 no-asm --prefix=${CMAKE_CURRENT_BINARY_DIR}/build"
+  COMMAND "${NMAKE_64_EXECUTABLE} -f ms/ntdll.mak"
```
Comment (devshgraphicsprogramming) [inline 3rdparty/openssl/CMakeLists.txt]: ntdll makes the shared lib, not the static that we want

**Comment by devshgraphicsprogramming** [comment]
@Crisspl you're getting linker errors on the examples because the static library of OpenSSL is not linked into the project because the build cmake command is not normal (dependency and location of the static library file not picked up automatically).

**Comment by Crisspl** [comment]
I'm not that stupid :P They're linked and I checked and the cmake code for this is in source/Irrlicht/CMakeLists

**Comment by devshgraphicsprogramming** [comment]
but wouldn't they need to be linked to every example as well?

**Comment by Crisspl** [comment]
The `INTERFACE` does that

**Comment by devshgraphicsprogramming** [comment]
hmm, ok I guess I'll shut up now

**Comment by devshgraphicsprogramming** [comment]
@Crisspl did you add the custom command to cmake that automatically updates the submodules?

P.S. What did you have to change in OpenSSL to make it compile?

**Comment by Crisspl** [comment]
>did you add the custom command to cmake that automatically updates the submodules?

No, and won't until this shit builds.
Actually IMO it's completely useless, but ofc you'll get it if you want

>What did you have to change in OpenSSL to make it compile?

https://github.com/openssl/openssl/blob/OpenSSL_1_0_2-stable/crypto/cversion.c#L96

It gave weird errors since `OPENSSLDIR` was some windows-like path like `D:\....` and so it needed string-quotes. So something like this needed:
```
#ifdef OPENSSLDIR
#define STRINGIFY__XD(s) #s
        return "OPENSSLDIR: \"" STRINGIFY__XD(OPENSSLDIR) "\"";
#undef STRINGIFY__XD
#else
        return "OPENSSLDIR: N/A";
#endif
```

**Comment by devshgraphicsprogramming** [comment]
Maybe `./configure` with `--openssldir="NA"` or set `--openssldir` identical to `--prefix`

P.S. Then no need to modify OpenSSL source.

**Comment by Crisspl** [comment]
Actually it needs string-quotes regardless of its value, i don't why it is like that in release version of openssl code

**Comment by devshgraphicsprogramming** [comment]
Maybe submit a pull request to them ;)

**Comment by Crisspl** [comment]
well, its hard to believe that it's not my fault in some way :P

**Comment by devshgraphicsprogramming** [comment]
> well, its hard to believe that it's not my fault in some way :P

Do it, at least they will tell you what you've done wrong.

**Comment by devshgraphicsprogramming** [comment]
About the git update submodules + git pull OpenSSL

https://stackoverflow.com/questions/15972898/cmake-how-to-run-a-add-custom-command-before-everything-else

**Comment by devshgraphicsprogramming** [comment]
TODO:
- [ ] Submit a PR to OpenSSL project
- [ ] IrrCompileConfig as CMAKE configured headers
- [ ] GCC Warnings (or at least make it build on Linux again)
- [x] CMake update submodules target
- [x] CMake pull latest OpenSSL 1.0.2 target

**Comment by Crisspl** [comment]
>About the git update submodules + git pull OpenSSL

I think it could/should be done with [execute_process](https://cmake.org/cmake/help/v3.12/command/execute_process.html) (executes before generation of makefiles/projects). It also makes sense since all the submodules must be there before cmake starts looking for their sources.

**Comment by devshgraphicsprogramming** [comment]
ok I might do it then, I'm fixing my master branch to compile on Linux right now

**Comment by devshgraphicsprogramming** [comment]
@Crisspl do you expect this to break on MSVC if I fetch the latest libshaderc ?

**Comment by Crisspl** [comment]
No, actually I think I have latest shaderc locally but not sure now

**Comment by devshgraphicsprogramming** [comment]
oh ok, because they fixed the `touch` thing on linux

**Comment by Crisspl** [comment]
oh, yea, you can just remove the line mentioning issue about this in root cmakelists.txt and that would be all I think. Anyway.. shouldn't break anything even if you'd left it

**Comment by devshgraphicsprogramming** [comment]
BTW, I got this beauty while configuring
```
CMake Warning (dev) at cmake/add_pch.cmake:63 (GET_DIRECTORY_PROPERTY):
  Policy CMP0059 is not set: Do not treat DEFINITIONS as a built-in directory
  property.  Run "cmake --help-policy CMP0059" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.
Call Stack (most recent call first):
  source/Irrlicht/CMakeLists.txt:272 (ADD_PRECOMPILED_HEADER)
This warning is for project developers.  Use -Wno-dev to suppress it.

/usr/bin/c++ -DPCHCOMPILE -O3;-DNDEBUG;;-DPNG_THREAD_UNSAFE_OK;-DPNG_NO_MMX_CODE;-DPNG_NO_MNG_FEATURES;-D_7ZIP_ST;-DSPIRV_CROSS_EXCEPTIONS_TO_ASSERTIONS;-DIRRLICHT_FAST_MATH -x c++-header -o {_output} REPOSITORY_ROOT/source/Irrlicht/pch.h
CMake Warning (dev) at cmake/add_pch.cmake:63 (GET_DIRECTORY_PROPERTY):
  Policy CMP0059 is not set: Do not treat DEFINITIONS as a built-in directory
  property.  Run "cmake --help-policy CMP0059" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.
Call Stack (most recent call first):
  source/Irrlicht/CMakeLists.txt:306 (ADD_PRECOMPILED_HEADER)
This warning is for project developers.  Use -Wno-dev to suppress it.

/usr/bin/c++ -DPCHCOMPILE -O3;-DNDEBUG;;-DPNG_THREAD_UNSAFE_OK;-DPNG_NO_MMX_CODE;-DPNG_NO_MNG_FEATURES;-D_7ZIP_ST;-DSPIRV_CROSS_EXCEPTIONS_TO_ASSERTIONS;-DIRRLICHT_FAST_MATH -x c++-header -o {_output} REPOSITORY_ROOT/source/Irrlicht/pch.h
```

obvs, `REPOSITORY_ROOT` is my full filesystme path

**Comment by Crisspl** [comment]
Just a rename DEFINITIONS->COMPILE_DEFINITIONS is needed. Apparently the pch script is a little old
https://cmake.org/cmake/help/v3.3/prop_dir/DEFINITIONS.html
https://cmake.org/cmake/help/v3.3/prop_dir/COMPILE_DEFINITIONS.html

**Comment by devshgraphicsprogramming** [comment]
awesome, then accept my pull request and do it ;)

Also my PCH on Linux can't find headers (errors out at `#include "aabbox3d.h"` when building the PCH) so Linux doesn't build

**Pull request merged**